core/domain/auth_jobs_one_off.py

```diff
@@ -0,0 +1,188 @@
+
+    @staticmethod
+    def map(user):
+        auth_id = firebase_auth_services.get_auth_id_from_user_id(user.id)
+        if auth_id is not None:
+            yield (POPULATED_KEY, 1)
+        else:
+            yield (
+                NOT_POPULATED_KEY,
+                (_strip_uid_prefix(user.id), user.id, user.email, user.deleted))
+
+    @staticmethod
+    def reduce(key, values):
+        if key == POPULATED_KEY:
+            yield (AUDIT_KEY, len(values))
+            return
+
+        try:
+            app = firebase_admin.initialize_app()
+        except Exception as exception:
+            yield (WARNING_KEY, repr(exception))
+            return
+
+        # NOTE: This is only sorted to make unit testing easier.
+        user_fields = sorted(ast.literal_eval(v) for v in values)
+        user_records = [
+            firebase_auth.ImportUserRecord(
+                uid=auth_id, email=email, disabled=disabled,
+                email_verified=True)
+            for auth_id, _, email, disabled in user_fields
+        ]
+
+        # The Firebase Admin SDK places a hard-limit on the number of users that
+        # can be "imported" in a single call. To compensate, we break up the
+        # users into chunks.
+        offsets = python_utils.RANGE(
+            0, len(user_records), FIREBASE_IMPORT_USERS_MAX_LEN)
+        results = (
+            _populate_firebase([record for record in record_group if record])
+            for record_group in _grouper(
+                user_records, FIREBASE_IMPORT_USERS_MAX_LEN))
+
+        assocs_to_create = []
+        for offset, (result, exception) in python_utils.ZIP(offsets, results):
+            if exception is not None:
+                yield (FAILURE_KEY, repr(exception))
+            else:
+                successful_indices = set(python_utils.RANGE(
+                    result.success_count + result.failure_count))
+                for error in result.errors:
+                    successful_indices.remove(error.index)
+                    debug_info = (
+                        'Import user_id=%r failed: %s' % (
+                            user_fields[offset + error.index][1], error.reason))
+                    yield (FAILURE_KEY, debug_info)
+                assocs_to_create.extend(
+                    auth_domain.AuthIdUserIdPair(*user_fields[offset + i][:2])
+                    for i in successful_indices)
+
+        if assocs_to_create:
+            firebase_auth_services.associate_multi_auth_ids_with_user_ids(
+                assocs_to_create)
+            yield (SUCCESS_KEY, len(assocs_to_create))
+
+        try:
+            firebase_admin.delete_app(app)
+        except Exception as exception:
+            yield (WARNING_KEY, repr(exception))
+
+
+def _grouper(iterable, chunk_len):
+    """Collect data into fixed-length chunks.
+
+    Source: https://docs.python.org/3/library/itertools.html#itertools-recipes.
+
+    Example:
+        grouper('ABCDEFG', 3, 'x') --> ABC DEF Gxx"
```

Comment:
I'm confused by this example. There are 3 args provided and only two in the function signature.

Reply:
Ah that's because I made some slight changes. Decided to undo them and use it exactly as written in the standard docs.